File: openhpi/trunk/baselibs/csharp/openhpi_baselib/Hpi.cs
/*      -*- c# -*-
 *
 * Copyright (C) 2012, Pigeon Point Systems
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  This
 * file and program are licensed under a BSD style license.  See
 * the Copying file included with the OpenHPI distribution for
 * full licensing terms.
 *
 * Author(s):
 *      Anton Pak <anton.pak@pigeonpoint.com>
 */

using System;


namespace org {
namespace openhpi {


/**********************************************************
 * HPI API
 * Specific functions that were hard to auto-generate.
 * So they were written manually.
 *********************************************************/
public static partial class Api
{
    public static long saHpiSessionOpen(
        long DomainId,
        out long SessionId,
        Object SecurityParams
    )
    {
        long rv;
        bool rc;

        SessionId = 0;

        if ( SecurityParams != null ) {
            return HpiConst.SA_ERR_HPI_INVALID_PARAMS;
        }

        HpiSession s = HpiCore.CreateSession( DomainId );
        if ( s == null ) {
            return HpiConst.SA_ERR_HPI_INVALID_DOMAIN;
        }
        OhpiMarshal m = s.GetMarshal();
        if ( m == null ) {
            HpiCore.RemoveSession( s );
            return HpiConst.SA_ERR_HPI_NO_RESPONSE;
        }

        m.MarshalSaHpiDomainIdT( s.GetRemoteDid() );
        rc = m.Interchange( OhpiConst.RPC_SAHPI_SESSION_OPEN );
        if ( !rc ) {
            m.Close();
            HpiCore.RemoveSession( s );
            return HpiConst.SA_ERR_HPI_NO_RESPONSE;
        }
        rv = m.DemarshalSaErrorT();
        if ( rv == HpiConst.SA_OK ) {
            SessionId = m.DemarshalSaHpiSessionIdT();
            s.SetRemoteSid( SessionId );
            SessionId = s.GetLocalSid();
        }
        s.PutMarshal( m );
        if ( rv != HpiConst.SA_OK ) {
            HpiCore.RemoveSession( s );
        }

        return rv;
    }

    public static long saHpiSessionClose(
        long SessionId
    )
    {
        long rv;
        bool rc;
        HpiSession s = HpiCore.GetSession( SessionId );
        if ( s == null ) {
            return HpiConst.SA_ERR_HPI_INVALID_SESSION;
        }
        OhpiMarshal m = s.GetMarshal();
        if ( m == null ) {
            return HpiConst.SA_ERR_HPI_NO_RESPONSE;
        }

        m.MarshalSaHpiSessionIdT( s.GetRemoteSid() );
        rc = m.Interchange( OhpiConst.RPC_SAHPI_SESSION_CLOSE );
        if ( !rc ) {
            m.Close();
            return HpiConst.SA_ERR_HPI_NO_RESPONSE;
        }
        rv = m.DemarshalSaErrorT();
        if ( rv == HpiConst.SA_OK ) {
            // No output arguments
        }
        s.PutMarshal( m );
        if ( rv == HpiConst.SA_OK ) {
            HpiCore.RemoveSession( s );
        }

        return rv;
    }

    public static long saHpiMyEntityPathGet(
        long SessionId,
        out SaHpiEntityPathT EntityPath
    )
    {
        EntityPath = null;

        HpiSession s = HpiCore.GetSession( SessionId );
        if ( s == null ) {
            return HpiConst.SA_ERR_HPI_INVALID_SESSION;
        }
        EntityPath = HpiCore.GetMyEntity();
        if ( EntityPath == null ) {
            return HpiConst.SA_ERR_HPI_UNKNOWN;
        }

        return HpiConst.SA_OK;
    }

    public static long saHpiDimiTestStart(
        long SessionId,
        long ResourceId,
        long DimiNum,
        long TestNum,
        long NumberOfParams,
        SaHpiDimiTestVariableParamsT[] ParamsList
    )
    {
        long rv;
        bool rc;
        if ( NumberOfParams != 0 ) {
            if ( ParamsList == null ) {
                return HpiConst.SA_ERR_HPI_INVALID_PARAMS;
            }
            if ( NumberOfParams > ParamsList.Length ) {
                return HpiConst.SA_ERR_HPI_INVALID_PARAMS;
            }
            for ( int i = 0; i < NumberOfParams; ++i ) {
                rc = HpiUtil.Check( ParamsList[i] );
                if ( !rc ) {
                    return HpiConst.SA_ERR_HPI_INVALID_PARAMS;
                }
            }
        }
        HpiSession s = HpiCore.GetSession( SessionId );
        if ( s == null ) {
            return HpiConst.SA_ERR_HPI_INVALID_SESSION;
        }
        OhpiMarshal m = s.GetMarshal();
        if ( m == null ) {
            return HpiConst.SA_ERR_HPI_NO_RESPONSE;
        }

        m.MarshalSaHpiSessionIdT( s.GetRemoteSid() );
        m.MarshalSaHpiResourceIdT( ResourceId );
        m.MarshalSaHpiDimiNumT( DimiNum );
        m.MarshalSaHpiDimiTestNumT( TestNum );
        m.MarshalSaHpiUint8T( NumberOfParams );
        for ( int i = 0; i < NumberOfParams; ++i ) {
            m.MarshalSaHpiDimiTestVariableParamsT( ParamsList[i] );
        }
        rc = m.Interchange( OhpiConst.RPC_SAHPI_DIMI_TEST_START );
        if ( !rc ) {
            m.Close();
            return HpiConst.SA_ERR_HPI_NO_RESPONSE;
        }
        rv = m.DemarshalSaErrorT();
        if ( rv == HpiConst.SA_OK ) {
            // No output arguments
        }
        s.PutMarshal( m );

        return rv;
    }

};


}; // namespace openhpi
}; // namespace org

File: openhpi/trunk/baselibs/python/openhpi_baselib/Hpi.py

#      -*- python -*-
#
# Copyright (C) 2012, Pigeon Point Systems
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  This
# file and program are licensed under a BSD style license.  See
# the Copying file included with the OpenHPI distribution for
# full licensing terms.
#
# Author(s):
#      Anton Pak <anton.pak@pigeonpoint.com>
#

from openhpi_baselib.HpiGen import *


#**********************************************************
# HPI API
# Specific functions that were hard to auto-generate.
# So they were written manually.
#**********************************************************

#**********************************************************
def saHpiSessionOpen(
    DomainId,
    SecurityParams
):
    if SecurityParams is not None:
        return ( SA_ERR_HPI_INVALID_PARAMS, None )
    s = HpiCore.createSession( DomainId )
    if s is None:
        return ( SA_ERR_HPI_INVALID_DOMAIN, None )
    m = s.getMarshal()
    if m is None:
        HpiCore.removeSession( s )
        return ( SA_ERR_HPI_NO_RESPONSE, None )

    m.marshalSaHpiDomainIdT( s.getRemoteDid() )
    rc = m.interchange( OhpiDataTypes.RPC_SAHPI_SESSION_OPEN )
    if not rc:
        m.close()
        HpiCore.removeSession( s )
        return ( SA_ERR_HPI_NO_RESPONSE, None )
    rv = m.demarshalSaErrorT()
    if rv == SA_OK:
        SessionId = m.demarshalSaHpiSessionIdT()
        s.setRemoteSid( SessionId )
        SessionId = s.getLocalSid()

    s.putMarshal( m )

    if rv != SA_OK:
        HpiCore.removeSession( s )
        return ( rv, None )
    return ( SA_OK, SessionId )

#**********************************************************
def saHpiSessionClose(
    SessionId
):
    s = HpiCore.getSession( SessionId )
    if s is None:
        return SA_ERR_HPI_INVALID_SESSION
    m = s.getMarshal()
    if m is None:
        return SA_ERR_HPI_NO_RESPONSE

    m.marshalSaHpiSessionIdT( s.getRemoteSid() )
    rc = m.interchange( OhpiDataTypes.RPC_SAHPI_SESSION_CLOSE )
    if not rc:
        m.close()
        return SA_ERR_HPI_NO_RESPONSE
    rv = m.demarshalSaErrorT()
    # No output arguments

    s.putMarshal( m )

    if rv == SA_OK:
        HpiCore.removeSession( s )
    if rv != SA_OK:
        return rv
    return SA_OK

#**********************************************************
def saHpiMyEntityPathGet(
    SessionId
):
    s = HpiCore.getSession( SessionId )
    if s is None:
        return ( SA_ERR_HPI_INVALID_SESSION, None )
    my_ep = HpiCore.getMyEntity()
    if my_ep is None:
        return ( SA_ERR_HPI_UNKNOWN, None )
    return ( SA_OK, my_ep )

#**********************************************************
def saHpiDimiTestStart(
    SessionId,
    ResourceId,
    DimiNum,
    TestNum,
    NumberOfParams,
    ParamsList
):
    if NumberOfParams != 0:
        if ParamsList is None:
            return SA_ERR_HPI_INVALID_PARAMS
        if not isinstance( ParamsList, list ):
            return SA_ERR_HPI_INVALID_PARAMS
        if NumberOfParams > len( ParamsList ):
            return SA_ERR_HPI_INVALID_PARAMS
        for p in ParamsList:
            rc = HpiUtil.checkSaHpiDimiTestVariableParamsT( p )
            if not rc:
                return SA_ERR_HPI_INVALID_PARAMS

    s = HpiCore.getSession( SessionId )
    if s is None:
        return SA_ERR_HPI_INVALID_SESSION
    m = s.getMarshal()
    if m is None:
        return SA_ERR_HPI_NO_RESPONSE

    m.marshalSaHpiSessionIdT( s.getRemoteSid() )
    m.marshalSaHpiResourceIdT( ResourceId )
    m.marshalSaHpiDimiNumT( DimiNum )
    m.marshalSaHpiDimiTestNumT( TestNum )
    m.marshalSaHpiUint8T( NumberOfParams )
    if ParamsList is not None:
        for p in ParamsList:
            m.marshalSaHpiDimiTestVariableParamsT( p )
    rc = m.interchange( OhpiDataTypes.RPC_SAHPI_DIMI_TEST_START )
    if not rc:
        m.close()
        return SA_ERR_HPI_NO_RESPONSE
    rv = m.demarshalSaErrorT()
    # No output arguments

    s.putMarshal( m )

    if rv != SA_OK:
        return rv
    return SA_OK

